src/screens/token-launcher/helpers/calculateTokenomics.ts

```diff
@@ -0,0 +1,132 @@
+import { AIRDROP_BPS, CREATOR_BPS, CREATOR_BPS_WITH_AIRDROP, RAINBOW_BPS } from '../constants';
+import { TokenLauncherSDK } from '@/hooks/useTokenLauncher';
+import { parseUnits } from '@ethersproject/units';
+// 1% fee
+const POOL_FEE = 10000;
+// 1_000_000 for 0.01% precision
+const FEE_DENOMINATOR = 1000000;
+
+export function calculateTokenomics({
+  targetMarketCapUsd,
+  totalSupply,
+  ethPriceUsd,
+  hasAirdrop = false,
+  // Creator's optional extra buy-in
+  amountInEth = 0,
+}: {
+  targetMarketCapUsd: number;
+  totalSupply: number;
+  ethPriceUsd: number;
+  hasAirdrop: boolean;
+  amountInEth?: number;
+}) {
+  if (totalSupply === 0) {
+    return undefined;
+  }
+  // Calculate supply allocations (same logic as in contract)
+  const creatorBaseBips = hasAirdrop ? CREATOR_BPS_WITH_AIRDROP : CREATOR_BPS;
+  const airdropAllocationBips = hasAirdrop ? AIRDROP_BPS : 0;
+  const rainbowFoundationAllocationBips = RAINBOW_BPS;
+  let lpAllocationBips = 10000 - creatorBaseBips - airdropAllocationBips - rainbowFoundationAllocationBips;
+  let creatorAllocationBips = creatorBaseBips;
+
+  let creatorAmount = (totalSupply * creatorBaseBips) / 10000;
+  const airdropAmount = (totalSupply * airdropAllocationBips) / 10000;
+  const rainbowFoundationAmount = (totalSupply * rainbowFoundationAllocationBips) / 10000;
+  let lpSupply = totalSupply - creatorAmount - airdropAmount - rainbowFoundationAmount;
+
+  // Calculate required price per token to achieve market cap
+  const targetMarketCapEth = targetMarketCapUsd / ethPriceUsd;
+  const targetPriceUsd = targetMarketCapUsd / totalSupply;
+  const targetPriceEth = targetPriceUsd / ethPriceUsd;
+
+  const tick = TokenLauncherSDK.getInitialTick(parseUnits(targetPriceEth?.toFixed(18) ?? '0', 18));
+
+  // Calculate actual starting values with rounded tick
+  let actualPriceEth = Math.pow(1.0001, tick);
+  let actualPriceUsd = actualPriceEth * ethPriceUsd;
+  let actualMarketCapUsd = actualPriceUsd * totalSupply;
+  let actualMarketCapEth = actualMarketCapUsd / ethPriceUsd;
+
+  // Calculate swap outcome if amountInEth is provided
+  let swap = undefined;
+
+  // TODO: all these calculations are wrong
```

Comment:
leaving a note here

Reply:
😅 this file has received updates since this comment was made